Kvikio backend entrypoint with Zarr v3

weiji14 opened the pull request:
Branched off the work at #10, this PR registers a `kvikio` backend that allows reading data from Zarr (v3) stores directly into GPU memory using NVIDIA GPU Direct Storage (GDS) without going through host (CPU) memory.

**Preview** demo at https://cupy-xarray--70.org.readthedocs.build/70/examples/07_kvikio.html

Towards enabling:

```python
ds = xr.open_dataset("/tmp/air-temp.zarr", engine="kvikio", consolidated=False)
assert isinstance(ds.air.data, cp.ndarray)
```

Requires some patches from un-released versions:
- [x] ~kvikio>=25.04.00a (https://github.com/rapidsai/kvikio/pull/646)~ Now in [`kvikio=25.04.00`](https://github.com/rapidsai/kvikio/releases/tag/v25.04.00)
- [x] ~xarray>=2025.1.3.dev22+g0184702f (https://github.com/pydata/xarray/pull/10078)~ Now in [`xarray=2025.03.0`](https://github.com/pydata/xarray/releases/tag/v2025.03.0)

Also requires:
- zarr >=3.0.3 (containing https://github.com/zarr-developers/zarr-python/pull/2751)

TODO:
- [x] Wait for stable release of kvikio>=25.04 and xarray>=2025.3.0
- [ ] Figure out where to insert `zarr.config.enable_gpu()` into the backend to read into `cupy.ndarray` by default?
- [ ] Write more unit tests
- [ ] Cleanup API and tutorial documentation

Limitations:
- Assumes that Zarr store is uncompressed. GPU-based decompression via nvCOMP is not supported yet (wait for https://github.com/zarr-developers/zarr-python/pull/2863)

References:
- https://docs.rapids.ai/api/kvikio/nightly/zarr/#zarr-python-3-x
- https://zarr.readthedocs.io/en/stable/user-guide/gpu.html#using-gpus-with-zarr

dcherian committed:
Add Kvikio backend entrypoint

dcherian committed:
Add demo notebook

dcherian committed:
Update kvikio notebook

dcherian committed:
Merge branch 'main' into kvikio-entrypoint

andersy005 committed:
Merge branch 'main' into kvikio-entrypoint

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

dcherian committed:
Update cupy_xarray/kvikio.py

andersy005 committed:
Merge branch 'main' into kvikio-entrypoint

dcherian committed:
Merge branch 'main' into kvikio-entrypoint

* main:
  Min xarray >= 0.19.0 (#25)
  Fix broken dask_array_type import (#24)

dcherian committed:
Add url, description.

dcherian committed:
Working

dcherian committed:
Updated notebook

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

dcherian committed:
Merge remote-tracking branch 'upstream/main' into kvikio-entrypoint

* upstream/main:
  Documentation Updates  📖 (#35)
  [pre-commit.ci] pre-commit autoupdate (#37)
  [pre-commit.ci] pre-commit autoupdate (#34)
  [pre-commit.ci] pre-commit autoupdate (#32)
  Update .pre-commit-config.yaml (#28)
  Expand installation doc (#27)

dcherian committed:
Add tests

weiji14 committed:
Merge branch 'main' into kvikio-entrypoint

weiji14 committed:
Move kvikio notebook under docs/source

Allow it to be rendered under the User Guide section.

weiji14 committed:
Add zarr as a dependency in ci/doc.yml

Will need it for the kvikio.zarr docs later.

weiji14 committed:
Add entry for KvikioBackendEntrypoint in API docs

Create new section in the API documentation page for the kvikIO engine. Added more docstrings to the kvikio.py file, and fixed some imports so things render nicely on the API page. Also added an intersphinx link to the kvikio docs at https://docs.rapids.ai/api/kvikio/stable.

weiji14 committed:
Fix input argument into CupyZarrArrayWrapper

Fixes error like `TypeError: ZarrArrayWrapper.__init__() takes 2 positional arguments but 3 were given`.

weiji14 committed:
Merge branch 'main' into kvikio-entrypoint

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

weiji14 committed:
Re-add kvikio backend entrypoint to pyproject.toml

Fix improper merge handling on d684dadc160f997ae7a6c72c73c61839d257f002

weiji14 committed:
Fix C408 and E402

weiji14 committed:
Use get_duck_array instead of get_array

Fix `TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly` on https://github.com/pydata/xarray/blob/v2024.11.0/xarray/core/indexing.py#L578

weiji14 committed:
Fix SIM108 Use ternary operator

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

weiji14 committed:
Install nightly version of kvikio=25.04.00a and zarr>=3.0.5

Need patches from https://github.com/rapidsai/kvikio/pull/646 and https://github.com/zarr-developers/zarr-python/pull/2751.

weiji14 committed:
Remove custom open_store_variable method from GDSZarrStore class

Can directly rely on upstream xarray's ZarrStore.open_store_variable method since Zarr v3 compatibility was added in https://github.com/pydata/xarray/pull/9552.

weiji14 committed:
Fix UserWarning compressor -> compressors

UserWarning: The `compressor` argument is deprecated. Use `compressors` instead.

weiji14 committed:
Reuse logic from xarray.backends.zarr.ZarrStore.open_group

Only difference is to pass the Zarr store's root filepath to kvikio.zarr.GDSStore instead of xarray.backends.zarr.ZarrStore.

weiji14 committed:
Install xarray=2025.1.3.dev22+g0184702f

Dev version containing patch at https://github.com/pydata/xarray/pull/10078 that fixes `TypeError: NumpyIndexingAdapter only wraps np.ndarray. Trying to wrap <class 'cupy.ndarray'>`.

weiji14 committed:
Add zarr.config.enable_gpu() context manager to test_lazy_indexing

Needed to return cupy.ndarray instead of numpy.ndarray objects. Should find a better place to put this `zarr.config.enable_gpu()` call later.

weiji14 committed:
Refresh kvikIO demo notebook

Rearranged some cells so the air-temperature.zarr store is created first. Added a couple of `zarr.config.enable_gpu()` statements in to ensure arrays are read to cupy.ndarray. Removed the flox section at the end.

weiji14 commented on cupy_xarray/tests/test_kvikio.py:
```diff
@@ -0,0 +1,54 @@
+
+@pytest.mark.parametrize("indexer", [slice(None), slice(2, 4), 2, [2, 3, 5]])
+def test_lazy_indexing(indexer, store):
+    with zarr.config.enable_gpu(), xr.open_dataset(store, engine="kvikio") as ds:
```
Ideally, `xr.open_dataset(store, engine="kvikio")` should have `zarr.config.enable_gpu()` set already, so GPU-backed cupy arrays are returned when a user accesses the arrays. Unless we expect there to be users who wants to use the kvikio engine while returning CPU-backed numpy arrays. Default should probably be cupy though?

TomAugspurger replied:
I'm not sure... The (v3) kvikio.zarr.GDSStore is written to accept either. I'm tempted to leave this up to the user (so make them configure Zarr correctly) rather than assuming they want to use GPU memory, but maybe `engine="kvikio"` is sufficient indication that they want stuff on the GPU.

weiji14 replied:
Yeah, `kvikio.zarr.GDSStore` can return numpy (CPU) arrays too, and I guess there might be a case where someone wants nvCOMP decompression on the GPU (once https://github.com/zarr-developers/zarr-python/pull/2863 is implemented), but have a numpy array returned in CPU memory? But I do feel that `engine="kvikio"` should default to putting things on GPU/cupy, while having a way to toggle it off as needed.

Looking at [`xr.open_dataset`](https://docs.xarray.dev/en/v2025.01.2/generated/xarray.open_dataset.html), I'm thinking if it might make sense to pass something into the `backend_kwargs` parameter to indicate the user's preference for GPU or CPU outputs. Is the `from_array_kwargs` or `chunked_array_type` parameters (experimental API) the intended use for this actually?

akshaysubr replied:
I am inclined to agree that `kvikio.zarr.GDSStore` should have the ability to not enable the GPU zarr configs even if someone wants to get data on the GPU since they can try and use a different `BufferPrototype` that return torch tensors instead of cupy arrays, for example.

But for ease of use, it might be better to default to `zarr.config.enable_gpu()` and that can be overridden by passing in a different context manager? So `xr.open_dataset(store, engine="kvikio")` will default to using cupy arrays, but `xr.open_dataset(store, engine="kvikio", zarr_context=contextlib.nullcontext())` would return numpy arrays and `xr.open_dataset(store, engine="kvikio", zarr_context=MyCustomTorchContext())` would return torch tensors?

weiji14 replied:
I think this is similar to the discussion at https://github.com/zarr-developers/zarr-python/issues/2473. That `zarr_context` parameter is exactly what I'm thinking of, I'm trying to find out if there's a way to implement this as is by setting/monkeypatching the default `prototype` in `kvikio.zarr.GDSStore`'s `get` function.

weiji14 committed:
Merge branch 'main' into kvikio-backend

jacobtomlinson commented:
cc @jakirkham @TomAugspurger @madsbk @ncclementi who may be interested in this

TomAugspurger reviewed (commented):
High-level question: is are these two functionally equivalent?

```python
ds = xr.open_dataset(kvikio.zarr.GDSStore("data.zarr"), engine="zarr")
ds = xr.open_dataset("data.zarr", engine="kvikio")
```

Do people have a preference on what to recommend to users?

weiji14 commented on cupy_xarray/kvikio.py:
```diff
@@ -0,0 +1,90 @@
+        filename_or_obj = _normalize_path(filename_or_obj)
+        if not store:
+            store = ZarrStore.open_group(
+                store=kvikio.zarr.GDSStore(root=filename_or_obj),
```
> High-level question: is are these two functionally equivalent?
> 
> ```python
> ds = xr.open_dataset(kvikio.zarr.GDSStore("data.zarr"), engine="zarr")
> ds = xr.open_dataset("data.zarr", engine="kvikio")
> ```
> 
> Do people have a preference on what to recommend to users?

Yes, those two are functionally the same. The `xr.open_dataset(kvikio.zarr.GDSStore("data.zarr"), engine="zarr")` style works today (with the xarray [patch](https://github.com/pydata/xarray/pull/10078) and kvikio=25.04.00a), while `xr.open_dataset("data.zarr", engine="kvikio")` is more just a convenience syntax. I was hoping to also sneak in the `zarr.config.enable_gpu()` in the backend somewhere to get `cupy` arrays by default (but we can discuss that in the https://github.com/xarray-contrib/cupy-xarray/pull/70/files#r1988251560 thread).

weiji14 committed:
Bump xarray from 2025.1.3.dev22+g0184702f to 2025.03.0

Bumps [xarray](https://github.com/pydata/xarray) from 2025.1.3.dev22+g0184702f to 2025.03.0.
- [Release notes](https://github.com/pydata/xarray/releases)
- [Changelog](https://github.com/pydata/xarray/blob/main/HOW_TO_RELEASE.md)
- [Commits](https://github.com/pydata/xarray/compare/0184702f16c3f744fc9096c7dac690626dcc6922...v2025.03.0)

Use stable version of xarray, also with patch https://github.com/pydata/xarray/pull/10081

jakirkham commented:
cc @akshaysubr

weiji14 committed:
Try overriding default prototype to be GPU buffer

Using functools.partial to override default buffer protocol to be GPU buffer instead of CPU buffer. Not quite working as expected, but hopefully gets a point across.

weiji14 committed:
Merge branch 'main' into kvikio-backend

weiji14 committed:
Update to stable version of kvikio=25.04.00

Stable release at https://github.com/rapidsai/kvikio/releases/tag/v25.04.00
